### DISCUSSION TITLE
Seemingly incorrect ground state energy for 2D J1-J2 model

### DISCUSSION THREAD

**Discussion by efrizoid** [QuSpin/QuSpin#630] [Q&A]
Greetings,

firstly, I want to express my appreciation for the creators and contributors of this library. It is very useful in my work!

I am currently researching neural network quantum states (NQSs) applied to 2D J1-J2 spin models and use QuSpin exact diagonalization (ED) to compute the ground states for comparison. Interestingly, during my experimentation with different values of the J2/J1 fraction for 6x6 lattice, I discovered that around J2/J1 = 1, the ground state energy obtained from NQS is consistently lower than that obtained from QuSpin. This is unexpected as the ground state obtained by ED should be more accurate than the one obtained from NQS.

The concrete energy values for 6x6 J2/J1 = 1:
- E_ed = -25.47
- E_nqs = -25.60

The difference, though small, significantly exceeds the confidence interval of the E_nqs estimation. The same happens with e.g. J2/J1 = 0.95

However, for other J2/J1 values such as 0, 0.25, 0.5, and 0.75, the energies match pretty well and E_ed is consistently smaller. This is true also in the case of a 4x4 lattice where even with J2/J1=1 E_ed is slightly smaller than E_nqs.

Since this behavior occurs within a narrow parameter range, it appears most likely that there is an issue with ED convergence to the true ground state. It would be helpful to hear any suggestions or ideas for enhancing the performance of ED in this particular scenario.

Below is the code that I use for ED. It is not completely reproducible since it imports my own Lattice class to get the indexes of nearest (_first_graph_edges_) and second nearest (_second_graph_edges_) neighbours. 

```python
import numpy as np
from quspin.operators import hamiltonian
from quspin.basis import spin_basis_general
from j1j2_lattice import Lattice


lattice = Lattice(
    size=6,
    J1=1 / 4,
    J2=1 / 4,
    periodic=True
)

s = np.arange(lattice.n_sites)  # sites [0,1,2,....]
x = s % lattice.size  # x positions for sites
y = s // lattice.size  # y positions for sites
T_x = (x + 1) % lattice.size + lattice.size * y  # translation along x-direction
T_y = x + lattice.size * ((y + 1) % lattice.size)  # translation along y-direction
P_x = x + lattice.size * (lattice.size - y - 1)  # reflection about x-axis
P_y = (lattice.size - x - 1) + lattice.size * y  # reflection about y-axis
Z = -(s + 1)  # spin inversion
D_x = s.reshape(lattice.size, lattice.size).T.flatten()  # reflection around diagonal from upper left to lower right

basis = spin_basis_general(
    lattice.n_sites,
    Nup=lattice.n_sites // 2,
    pauli=1,
    kxblock=(T_x, 0),
    kyblock=(T_y, 0),
    pxblock=(P_x, 0),
    pyblock=(P_y, 0),
    zblock=(Z, 0),
    dxblock=(D_x, 0),
)

coupling_list = []

for edge in np.asarray(lattice.first_graph_edges):
    j, k = edge
    coupling_list.append([lattice.J1, j, k])

for edge in np.asarray(lattice.second_graph_edges):
    j, k = edge
    coupling_list.append([lattice.J2, j, k])

static = [["xx", coupling_list], ["yy", coupling_list], ["zz", coupling_list]]
H = hamiltonian(static, [], basis=basis, check_symm=False, check_herm=False, check_pcon=False, dtype=np.complex128).as_sparse_format()
E_gs, psi_gs = H.eigsh(k=1, which='SA')
print(f"E_ed = {E_gs[0]}")
```

**Comment by mgbukov** [comment]
Interesting! Have you tried running some DMRG as a tie-breaker? E.g., tenpy
shouldn't be difficult to set up:

 https://tenpy.readthedocs.io/en/latest/

Let me know what comes out!

Btw, we used quspin for the J1-J2 model some time ago. We were focusing on
J2=J1/2 point, but I remember there were some tricks to speed up the
calculation of the basis on a 6x6 lattice (like the order of blocks would
matter in this case and it can save you time if you do it properly). Here
how I did it back at the time:

sites = np.arange(self.N_sites,dtype=np.int32) # sites [0,1,2,....]

x = sites%Lx # x positions for sites
y = sites//Lx # y positions for sites

T_x = (x+1)%Lx + Lx*y # translation along x-direction
T_y = x +Lx*((y+1)%Ly) # translation along y-direction

P_x = x + Lx*(Ly-y-1) # reflection about x-axis
P_y = (Lx-x-1) + Lx*y # reflection about y-axis
P_d = y + Lx*x

Z   = -(sites+1) # spin inversion

###### setting up bases ######
self.basis_symm = spin_basis_general(self.N_sites, pauli=False,
Ns_block_est=200,
Nup=self.N_sites//2,
kxblock=(T_x,0),kyblock=(T_y,0),
pdblock=(P_d,0),
pxblock=(P_x,0),pyblock=(P_y,0),
zblock=(Z,0),
block_order=['zblock','pdblock','pyblock','pxblock','kyblock','kxblock']
)
self.basis = spin_basis_general(self.N_sites, pauli=False,
Nup=self.N_sites//2)






On Thu, Jun 29, 2023 at 10:05 AM EimantasLedinauskas <
***@***.***> wrote:

> Greetings,
>
> firstly, I want to express my appreciation for the creators and
> contributors of this library. It is very useful in my work!
>
> I am currently researching neural network quantum states (NQSs) applied to
> 2D J1-J2 spin models and use QuSpin exact diagonalization (ED) to compute
> the ground states for comparison. Interestingly, during my experimentation
> with different values of the J2/J1 fraction for 6x6 lattice, I discovered
> that around J2/J1 = 1, the ground state energy obtained from NQS is
> consistently lower than that obtained from QuSpin. This is unexpected as
> the ground state obtained by ED should be more accurate than the one
> obtained from NQS.
>
> The concrete energy values for 6x6 J2/J1 = 1:
>
>    - E_ed = -25.47
>    - E_nqs = -25.60
>
> The difference, though small, significantly exceeds the confidence
> interval of the E_nqs estimation. The same happens with e.g. J2/J1 = 0.95
>
> However, for other J2/J1 values such as 0, 0.25, 0.5, and 0.75, the
> energies match pretty well and E_ed is consistently smaller. This is true
> also in the case of a 4x4 lattice where even with J2/J1=1 E_ed is slightly
> smaller than E_nqs.
>
> Since this behavior occurs within a narrow parameter range, it appears
> most likely that there is an issue with ED convergence to the true ground
> state. It would be helpful to hear any suggestions or ideas for enhancing
> the performance of ED in this particular scenario.
>
> Below is the code that I use for ED. It is not completely reproducible
> since it imports my own Lattice class to get the indexes of nearest (
> *first_graph_edges*) and second nearest (*second_graph_edges*) neighbours.
>
> import numpy as npfrom quspin.operators import hamiltonianfrom quspin.basis import spin_basis_generalfrom j1j2_lattice import Lattice
>
> lattice = Lattice(
>     size=6,
>     J1=1 / 4,
>     J2=1 / 4,
>     periodic=True
> )
> s = np.arange(lattice.n_sites)  # sites [0,1,2,....]x = s % lattice.size  # x positions for sitesy = s // lattice.size  # y positions for sitesT_x = (x + 1) % lattice.size + lattice.size * y  # translation along x-directionT_y = x + lattice.size * ((y + 1) % lattice.size)  # translation along y-directionP_x = x + lattice.size * (lattice.size - y - 1)  # reflection about x-axisP_y = (lattice.size - x - 1) + lattice.size * y  # reflection about y-axisZ = -(s + 1)  # spin inversionD_x = s.reshape(lattice.size, lattice.size).T.flatten()  # reflection around diagonal from upper left to lower right
> basis = spin_basis_general(
>     lattice.n_sites,
>     Nup=lattice.n_sites // 2,
>     pauli=1,
>     kxblock=(T_x, 0),
>     kyblock=(T_y, 0),
>     pxblock=(P_x, 0),
>     pyblock=(P_y, 0),
>     zblock=(Z, 0),
>     dxblock=(D_x, 0),
> )
> coupling_list = []
> for edge in np.asarray(lattice.first_graph_edges):
>     j, k = edge
>     coupling_list.append([lattice.J1, j, k])
> for edge in np.asarray(lattice.second_graph_edges):
>     j, k = edge
>     coupling_list.append([lattice.J2, j, k])
> static = [["xx", coupling_list], ["yy", coupling_list], ["zz", coupling_list]]H = hamiltonian(static, [], basis=basis, check_symm=False, check_herm=False, check_pcon=False, dtype=np.complex128).as_sparse_format()E_gs, psi_gs = H.eigsh(k=1, which='SA')print(f"E_ed = {E_gs[0]}")
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/QuSpin/QuSpin/discussions/630>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADHSUBXTOSSX3SC6EY7D2VDXNUZLRANCNFSM6AAAAAAZYEXPGM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Marin Bukov, PhD
Research Group Leader
Max Planck Institute for the Physics of Complex Systems
Noethnitzer Str. 38
Dresden 01187, Germany

url: https://www.pks.mpg.de/nqd
<http://quantum-dynamics.phys.uni-sofia.bg/>

**Reply by efrizoid** [reply]
Thanks for sharing the example code! Regarding the DMRG, I actually already tried running it. In 6x6 J2/J1 = 1 case, it converges to ~ -24.8 which is higher than both ED and NQS results.

**Comment by mgbukov** [comment]
very interesting! Did you play with the parameters of the mixer?

OK, back to ED: are you using scipy.eigsh(k=1)? Did you play with the other
paramters, like maxiter, etc.? It might make sense to see if using another
solver can help. I've heard good things about slepc and it looks like they
now have a python interface:
https://slepc.upv.es/slepc4py-current/docs/usrman/index.html.

On Thu, Jun 29, 2023 at 11:33 AM EimantasLedinauskas <
***@***.***> wrote:

> Thanks for sharing the example code! Regarding the DMRG, I actually
> already tried running it. In 6x6 J2/J1 = 1 case, it converges to ~ -24.8
> which is higher than both ED and NQS results.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/QuSpin/QuSpin/discussions/630#discussioncomment-6313522>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADHSUBX5Z5EIUXSPEJ67WBLXNVDX3ANCNFSM6AAAAAAZYEXPGM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Marin Bukov, PhD
Research Group Leader
Max Planck Institute for the Physics of Complex Systems
Noethnitzer Str. 38
Dresden 01187, Germany

url: https://www.pks.mpg.de/nqd
<http://quantum-dynamics.phys.uni-sofia.bg/>

**Reply by efrizoid** [reply]
After experimenting with DMRG parameters, I finally also obtained a lower energy than that achieved by ED. Specifically, -25.58, which closely aligns with the NQS value. So it appears that there is indeed an issue with ED in this case.

Regarding the ED solver, I use `eigsh(k=1, which='SA')`. I didn't try changing maxiter yet since it appears that convergence is achieved prior to reaching the maximum number of iterations (according to documentation it should raise ArpackNoConvergence exception otherwise).

I will attempt to experiment with other eigsh parameters and maybe also with SLEPc (although their python package documentation seems pretty scarce :) ).
